Add zperf loopback

krish2718 opened the pull request:
For testing Zephyr's network stack TX path.

Commit:
drivers: net: loopback: Make MTU configurable

This is useful to test higher packet sizes to simulate L2
Ethernet/Wi-Fi.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
drivers: net: loopback: Fix typo

Fix a typo in the comment.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

Commit:
samples: net: zperf: Add loopback overlay

This is useful to profile networking stack using loopback interface, a
TX only option is also added but disabled by default.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

jukkar commented on drivers/net/loopback.c:
```diff
@@ -89,6 +89,10 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
 
 	ARG_UNUSED(dev);
 
+#ifdef CONFIG_NET_LOOPBACK_DROP_ALL_PACKETS
+	return 0;
+#endif
+
```
Does not make any difference here, but I would suggest that we try to use the `IS_ENABLED()` macro when possible. This way perhaps the use gets more widespread and becomes a habit.
```
if (IS_ENABLED(CONFIG_NET_LOOPBACK_DROP_ALL_PACKETS)) {
     return 0;
}
```

rlubos commented on drivers/net/Kconfig:
```diff
@@ -191,6 +191,13 @@ config NET_LOOPBACK_MTU
 	help
 	  This option sets the MTU for loopback interface.
 
+config NET_LOOPBACK_DROP_ALL_PACKETS
```
There's already `config NET_LOOPBACK_SIMULATE_PACKET_DROP`, isn't this new config duplicating functionality? The application can set drop ratio to `1.0` with `loopback_set_packet_drop_ratio()`.

krish2718 replied:
True, but it also simulates real driver by adding clone, so, to avoid clone impacting the network stack bench mark for which this PR is used, added a new option.

krish2718 replied:
Moving that before clone, won't it change the current behaviour? Let me run some tests to see the performance impact, if its minimal then will use the existing define itself.

rlubos replied:
Well, the cloned packet is just dropped after being cloned, so skipping the cloning should not break anything.

krish2718 replied:
Sure, but the big comment says this is to simulate the real drivers (probably adds latency)

rlubos replied:
That comment is not related to the packet drop functionality, it was added when the wrong loopback behavior was fixed (previously the original packet was forwarded back to the RX path, instead of a clone: https://github.com/zephyrproject-rtos/zephyr/commit/0436ad9e78de4c5d7a3d820e95778523d9c1be78).

krish2718 replied:
Ah makes sense now, thanks. Will remove the new define.

rlubos commented on samples/net/zperf/overlay-loopback.conf:
```diff
@@ -0,0 +1,13 @@
+CONFIG_NET_PKT_TX_COUNT=48
+CONFIG_NET_BUF_RX_COUNT=32
+CONFIG_NET_BUF_TX_COUNT=96
+CONFIG_HEAP_MEM_POOL_SIZE=200000
```
Why heap? The sample does not use any heap at all.

krish2718 replied:
copy-paste, will remove.

Commit:
drivers: net: loopback: Optimize packet drop

If we are dropping packets, then drop then early without the clone, this
improves zperf performance.

Signed-off-by: Krishna T <krishna.t@nordicsemi.no>

krish2718 commented:
Apologies, not sure how I requested review of all, I don't even have permissions to add reviewers :-(

rlubos commented on drivers/net/loopback.c:
```diff
@@ -89,6 +89,19 @@ static int loopback_send(const struct device *dev, struct net_pkt *pkt)
+		/* Drop packets based on the loopback_packet_drop_ratio
+		 * a ratio of 0.2 will drop one every 5 packets
+		 */
+		loopback_packet_drop_state += loopback_packet_drop_ratio;
+		if (loopback_packet_drop_state >= 1.0f) {
+			/* Administrate we dropped a packet */
+			loopback_packet_drop_state -= 1.0f;
+			loopback_packet_dropped_count++;
```
Those variables are hidden under `#ifdef CONFIG_NET_LOOPBACK_SIMULATE_PACKET_DROP`, hence build failures in the CI. We need to either move them out, or stick to the `#ifdef CONFIG_NET_LOOPBACK_SIMULATE_PACKET_DROP` here as well.

rlubos reviewed (approved):
LGTM
